pkg/parse/root_sync_client.go

```diff
@@ -197,15 +204,32 @@ func (p *rootSyncStatusClient) SetSourceAnnotations(ctx context.Context, commit
 	err := opts.Client.Patch(ctx, rs,
 		client.RawPatch(types.MergePatchType, []byte(patch)),
 		client.FieldOwner(configsync.FieldManager))
-
 	if err != nil {
-		return fmt.Errorf("failed to patch RootSync annotations: %w\nPatch content: %s", err, patch)
+		// If denied by a webhook, assume it's for OCI signature validation, and therefore a source error.
+		if isRequestDeniedByWebhookStatusError(err) {
+			return status.SourceError.Wrap(err).Sprintf("webhook blocked patching RootSync annotation: %s", metadata.ImageToSyncAnnotationKey).Build()
+		}
+		return status.APIServerErrorf(err, "failed to patch RootSync annotation: %s", metadata.ImageToSyncAnnotationKey)
 	}
-
 	return nil
 }
 
-func (p *rootSyncStatusClient) SetRequiresRendering(ctx context.Context, renderingRequired bool) error {
+// isRequestDeniedByWebhookStatusError returns true if the error is a Kubernetes
+// *apierrors.StatusError
+func isRequestDeniedByWebhookStatusError(err error) bool {
+	if err == nil {
+		return false
+	}
+	var statusErr *apierrors.StatusError
+	if !errors.As(err, &statusErr) {
+		return false
+	}
+	// TODO: check for a specific error code to distinguish source errors, like
+	// OCI image verification failure.
+	return requestDeniedByWebhookPattern.MatchString(statusErr.Error())
```

Comment:
It is not reliable to determine the error type based on the error message. If the error message is changed in the future, we have to deal with multiple versions of strings.

Reply:
The only way to know for sure that a webhook rejected the error because of a source error is to have an interface schema or contact in place for the webhook to distinguish itself from other webhooks. Thats why I suggested using a specific HTTP status code, like 430, which is otherwise unused by HTTP and could be interpreted as a source error.

Alternatively we could require the webhook to send a specific error message and detect that. 

What I put here is a compromise to avoid needing a more specific interface contract. But as you mentioned, it's not perfect.

Reply:
Another alternative would be to just treat all these patch errors as APIServer errors, to be consistent with the other status update methods.

It's much more likely that our users will hit a network error here than a webhook rejection due to source, especially since we only have a couple customers who have asked for the signature verification feature.

Reply:
I think I'm okay with returning API errors when validation fails in the webhook server.

Reply:
I'm fine with just returning a 2002 APIServer error for now.

It has the downside that the user won't be able to detect that it was blocked by their webhook without doing error text matching, but I guess no customers have asked for that yet, and if no one's depending on it we can freely change the error type later, if we want.

Reply:
I updated the PR to remove the regex and source error. It's all APIServer errors now, unless the InternalError case happens, which it shouldn't.

Reply:
LGTM